File: packages/langchain-python/safestop_langchain/middleware.py
import json
from pathlib import Path
from typing import Any, Union

from .validator import validate_artifact


def with_safestop(agent: Any, config: dict) -> Any:
    """Wrap an agent with Safestop: validate artifact at call time; raise if invalid or not deployable.

    config:
        artifact: path to JSON file or artifact dict
    """
    artifact_cfg = config.get("artifact")
    if isinstance(artifact_cfg, (str, Path)):
        with open(artifact_cfg, encoding="utf-8") as f:
            artifact = json.load(f)
    else:
        artifact = artifact_cfg

    result = validate_artifact(artifact)
    if not result["valid"]:
        errors = "; ".join(result.get("errors", ["validation failed"]))
        raise ValueError(f"Safestop artifact invalid: {errors}")
    if not result.get("deployable"):
        missing = ", ".join(result.get("missing_requirements", []))
        raise ValueError(f"Safestop artifact not deployable: {missing}")

    return agent

File: packages/langchain-js/src/middleware.ts
import type { Runnable } from "@langchain/core/runnables";
import { validateArtifact, type SafestopArtifact } from "@safestop/core";
import { readFileSync } from "fs";

export interface SafestopConfig {
  artifact: string | SafestopArtifact;
  enforceQuantifiedLimits?: boolean;
  logReceiptsTo?: string;
}

/**
 * Wraps an agent (or any Runnable) with Safestop: loads and validates the governance
 * artifact at construction time. If invalid or not deployable, throws.
 * Runtime enforcement of quantified limits can be added in a future version.
 */
export function withSafestop<T extends Runnable>(agent: T, config: SafestopConfig): T {
  const artifact: SafestopArtifact =
    typeof config.artifact === "string"
      ? (JSON.parse(readFileSync(config.artifact, "utf-8")) as SafestopArtifact)
      : config.artifact;

  const validation = validateArtifact(artifact);
  if (!validation.valid) {
    const messages = validation.errors?.map((e) => `${e.instancePath}: ${e.message}`).join("; ");
    throw new Error(`Safestop artifact invalid: ${messages ?? "validation failed"}`);
  }
  if (!validation.deployable) {
    const missing = validation.missingRequirements?.join(", ") ?? "not deployable";
    throw new Error(`Safestop artifact not deployable: ${missing}`);
  }

  return agent;
}
